# Basic Pandas usage

In [2]:
import pandas as pd

## Introduction to Pandas

Pandas is a Python library to edit, load, visualize, clean, and manipulates data in general.

In [3]:
data = [[4, 2, 1],
        [3, 0, 1],
        [1, 0, 0]]

columns = ["apples", "bananas", "oranges"]
index = ["Monday", "Tuesday", "Wednesday"]

df = pd.DataFrame(data, index, columns)

print(df)

           apples  bananas  oranges
Monday          4        2        1
Tuesday         3        0        1
Wednesday       1        0        0


## Loading data into Pandas

This library is flexible allowing you to work with different popular data formats.

### Create a Pandas DataFrame from an online CSV

In [7]:
csv_url = "https://raw.githubusercontent.com/paiml/wine-ratings/main/wine-ratings.csv"

df  = pd.read_csv(csv_url, index_col = 0)

df.head(5)

name  grape  \
0  1000 Stories Bourbon Barrel Aged Batch Blue Ca...    NaN   
1  1000 Stories Bourbon Barrel Aged Gold Rush Red...    NaN   
2  1000 Stories Bourbon Barrel Aged Gold Rush Red...    NaN   
3    1000 Stories Bourbon Barrel Aged Zinfandel 2013    NaN   
4    1000 Stories Bourbon Barrel Aged Zinfandel 2014    NaN   

                    region   variety  rating  \
0    Mendocino, California  Red Wine    91.0   
1               California  Red Wine    89.0   
2               California  Red Wine    90.0   
3  North Coast, California  Red Wine    91.0   
4               California  Red Wine    90.0   

                                               notes  
0  This is a very special, limited release of 100...  
1  The California Gold Rush was a period of coura...  
2  The California Gold Rush was a period of coura...  
3  The wine has a deep, rich purple color. An int...  
4  Batch #004 is the first release of the 2014 vi...

### Load a CSV from a local file

In [8]:
df = pd.read_csv("./Data/world-championship-qualifier.csv")

df.head(5)

Rank             Name     Nationality Result Notes Group
0   1.0    Svatoslav Ton  Czech Republic   2.14     q     A
1   1.0     Toni Huikuri        Finlandi   2.14     q     A
2   1.0   James Brierley  United Kingdom   2.14     q     A
3   1.0    Noriyasu Arai           Japan   2.14     q     A
4   5.0  Yannick Tregaro          Sweden   2.14     q     A

### Load JSON from a local file

In [10]:
df = pd.read_json("./Data/world-championship-qualifier.json")

df.head(5)

Rank             Name     Nationality Result Notes Group
0   1.0    Svatoslav Ton  Czech Republic   2.14     q     A
1   1.0     Toni Huikuri        Finlandi   2.14     q     A
2   1.0   James Brierley  United Kingdom   2.14     q     A
3   1.0    Noriyasu Arai           Japan   2.14     q     A
4   5.0  Yannick Tregaro          Sweden   2.14     q     A

## Writing data from Pandas dataframes

Once data is loaded in a <span style="color:#CC0000">Dataframe</span>  object in Pandas, you have the ability to transform that data to various different formats. Most destinations will require a path on the filesystem for writing the output.

In [4]:
df = pd.read_csv("./Data/world-championship-qualifier.csv")
df

Rank                    Name     Nationality      Result    Notes Group
0    1.0           Svatoslav Ton  Czech Republic        2.14        q     A
1    1.0            Toni Huikuri        Finlandi        2.14        q     A
2    1.0          James Brierley  United Kingdom        2.14        q     A
3    1.0           Noriyasu Arai           Japan        2.14        q     A
4    5.0         Yannick Tregaro          Sweden        2.14        q     A
5    5.0       Dejan Vreljakovic              FR  Yugoslavia  2.14\tq     A
6    7.0            Alfredo Deza            Peru        2.10      NaN     A
7    8.0         Vagner Principe          Brazil        2.10      NaN     A
8    9.0  Alberto Juantorena Jr.            Cuba        2.10      NaN     A
9   10.0         Marcin Kaczocha          Poland        2.10      NaN     A
10  11.0         Andrey Krasulya         Ukraine        2.05      NaN     A
11  12.0            David Larsen   United States        2.05      NaN     A
12  13.0          Ronald Garlett       Australia        2.00      NaN     A
13  13.0           Oleg Prokopov         Belarus        2.00      NaN     A
14  15.0         Felipe Apablaza           Chile        2.00      NaN     A
15  16.0     Luis Soto Caballero     Puerto Rico        2.00      NaN     A
16   NaN            Zoltán Akacz         Hungary          NH      NaN     A
17   1.0            Mark Boswell          Canada        2.14        q     B
18   1.0          Ben Challenger  United Kingdom        2.14        q     B
19   1.0            Roman Fricke         Germany        2.14        q     B
20   1.0          Tivadar Kovács         Hungary        2.14        q     B
21   1.0             Dave Furman   United States        2.14        q     B
22   6.0      François Potgieter    South Africa        2.14        q     B
23   7.0             Sauli Niemi         Finland        2.10      NaN     B
24   7.0    Katsuyoshi Miyamichi           Japan        2.10      NaN     B
25   9.0           Marat Rakipov          Russia        2.10      NaN     B
26  10.0              Adi Mordel          Israel        2.10      NaN     B
27  11.0         Fabrício Romero          Brazil        2.10      NaN     B
28  12.0      Abderahmane Hammad         Algeria        2.05      NaN     B
29  12.0              Luke Temme       Australia        2.05      NaN     B
30  14.0       Aleksey Lesnichiy         Belarus        2.05      NaN     B
31  15.0            Ha Chung-Soo     South Korea        2.00      NaN     B
32  16.0           Dejan Dokleja         Croatia        2.00      NaN     B
33   NaN           Fawzi Warsame         Somalia          NH      NaN     B

The destinations are from a <span style="color:#CC0000">Dataframe</span> object itself, not from the <span style="color:#CC0000">pd</span>  object.

### Export a dataset to HTML

In [5]:
df.to_html("dataset.html")

The flexibility of Pandas to allow you to read and write to many different formats and destinations can be used to create quick utilities like transforming a CSV file to paste to Excel or Markdown.

### To Markdown and then to clipboard

In [7]:
from pandas.io.clipboards import to_clipboard

In [9]:
md = df.to_markdown()
to_clipboard(md, excel = False)

|    |   Rank | Name                   | Nationality    | Result     | Notes   | Group   |
|---:|-------:|:-----------------------|:---------------|:-----------|:--------|:--------|
|  0 |      1 | Svatoslav Ton          | Czech Republic | 2.14       | q       | A       |
|  1 |      1 | Toni Huikuri           | Finlandi       | 2.14       | q       | A       |
|  2 |      1 | James Brierley         | United Kingdom | 2.14       | q       | A       |
|  3 |      1 | Noriyasu Arai          | Japan          | 2.14       | q       | A       |
|  4 |      5 | Yannick Tregaro        | Sweden         | 2.14       | q       | A       |
|  5 |      5 | Dejan Vreljakovic      | FR             | Yugoslavia | 2.14	q         | A       |
|  6 |      7 | Alfredo Deza           | Peru           | 2.10       | nan     | A       |
|  7 |      8 | Vagner Principe        | Brazil         | 2.10       | nan     | A       |
|  8 |      9 | Alberto Juantorena Jr. | Cuba           | 2.10       | nan     | A       |
|  9 |     10 | Marcin Kaczocha        | Poland         | 2.10       | nan     | A       |
| 10 |     11 | Andrey Krasulya        | Ukraine        | 2.05       | nan     | A       |
| 11 |     12 | David Larsen           | United States  | 2.05       | nan     | A       |
| 12 |     13 | Ronald Garlett         | Australia      | 2.00       | nan     | A       |
| 13 |     13 | Oleg Prokopov          | Belarus        | 2.00       | nan     | A       |
| 14 |     15 | Felipe Apablaza        | Chile          | 2.00       | nan     | A       |
| 15 |     16 | Luis Soto Caballero    | Puerto Rico    | 2.00       | nan     | A       |
| 16 |    nan | Zoltán Akacz           | Hungary        | NH         | nan     | A       |
| 17 |      1 | Mark Boswell           | Canada         | 2.14       | q       | B       |
| 18 |      1 | Ben Challenger         | United Kingdom | 2.14       | q       | B       |
| 19 |      1 | Roman Fricke           | Germany        | 2.14       | q       | B       |
| 20 |      1 | Tivadar Kovács         | Hungary        | 2.14       | q       | B       |
| 21 |      1 | Dave Furman            | United States  | 2.14       | q       | B       |
| 22 |      6 | François Potgieter     | South Africa   | 2.14       | q       | B       |
| 23 |      7 | Sauli Niemi            | Finland        | 2.10       | nan     | B       |
| 24 |      7 | Katsuyoshi Miyamichi   | Japan          | 2.10       | nan     | B       |
| 25 |      9 | Marat Rakipov          | Russia         | 2.10       | nan     | B       |
| 26 |     10 | Adi Mordel             | Israel         | 2.10       | nan     | B       |
| 27 |     11 | Fabrício Romero        | Brazil         | 2.10       | nan     | B       |
| 28 |     12 | Abderahmane Hammad     | Algeria        | 2.05       | nan     | B       |
| 29 |     12 | Luke Temme             | Australia      | 2.05       | nan     | B       |
| 30 |     14 | Aleksey Lesnichiy      | Belarus        | 2.05       | nan     | B       |
| 31 |     15 | Ha Chung-Soo           | South Korea    | 2.00       | nan     | B       |
| 32 |     16 | Dejan Dokleja          | Croatia        | 2.00       | nan     | B       |
| 33 |    nan | Fawzi Warsame          | Somalia        | NH         | nan     | B       |

## Exploratory analysis with Pandas

One of the main tasks to perform with Pandas is exploratory analysis. Looking at data, finding  what is useful or potentially wrong with it so that you can clean it up are core practices of a data scientist and data engineer.

In [12]:
csv_url = "https://raw.githubusercontent.com/paiml/wine-ratings/main/wine-ratings.csv"
df  = pd.read_csv(csv_url, index_col = 0)

In [11]:
df.head(15)

name  grape  \
0   1000 Stories Bourbon Barrel Aged Batch Blue Ca...    NaN   
1   1000 Stories Bourbon Barrel Aged Gold Rush Red...    NaN   
2   1000 Stories Bourbon Barrel Aged Gold Rush Red...    NaN   
3     1000 Stories Bourbon Barrel Aged Zinfandel 2013    NaN   
4     1000 Stories Bourbon Barrel Aged Zinfandel 2014    NaN   
5     1000 Stories Bourbon Barrel Aged Zinfandel 2016    NaN   
6     1000 Stories Bourbon Barrel Aged Zinfandel 2017    NaN   
7                             12 Linajes Crianza 2014    NaN   
8                             12 Linajes Reserva 2012    NaN   
9                    14 Hands Cabernet Sauvignon 2010    NaN   
10                   14 Hands Cabernet Sauvignon 2011    NaN   
11                   14 Hands Cabernet Sauvignon 2015    NaN   
12                           14 Hands Chardonnay 2011    NaN   
13                14 Hands Hot to Trot Red Blend 2010    NaN   
14                14 Hands Hot to Trot Red Blend 2012    NaN   

                         region     variety  rating  \
0         Mendocino, California    Red Wine    91.0   
1                    California    Red Wine    89.0   
2                    California    Red Wine    90.0   
3       North Coast, California    Red Wine    91.0   
4                    California    Red Wine    90.0   
5                    California    Red Wine    91.0   
6                    California    Red Wine    92.0   
7       Ribera del Duero, Spain    Red Wine    92.0   
8       Ribera del Duero, Spain    Red Wine    94.0   
9   Columbia Valley, Washington    Red Wine    87.0   
10  Columbia Valley, Washington    Red Wine    89.0   
11  Columbia Valley, Washington    Red Wine    89.0   
12                   Washington  White Wine    89.0   
13                   Washington    Red Wine    88.0   
14                   Washington    Red Wine    88.0   

                                                notes  
0   This is a very special, limited release of 100...  
1   The California Gold Rush was a period of coura...  
2   The California Gold Rush was a period of coura...  
3   The wine has a deep, rich purple color. An int...  
4   Batch #004 is the first release of the 2014 vi...  
5   1,000 Stories Bourbon barrel-aged Zinfandel is...  
6   Batch 55 embodies an opulent vintage, which sa...  
7   Red with violet hues. The aromas are very inte...  
8   On the nose, a complex predominance of mineral...  
9   Concentrated aromas of dark stone fruits and t...  
10  Concentrated aromas of dark stone fruits and t...  
11  The 14 Hands Cabernet Sauvignon is a rich, jui...  
12  This Chardonnay displays aromas and flavors of...  
13  This Washington state red blend opens with gen...  
14  Flavors of ripe red fruit, supported by a fram...

In [14]:
# Description of the data
df.describe()

grape        rating
count    0.0  32780.000000
mean     NaN     91.186608
std      NaN      2.190391
min      NaN     85.000000
25%      NaN     90.000000
50%      NaN     91.000000
75%      NaN     92.000000
max      NaN     99.000000

In [15]:
# Metadata about the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32780 entries, 0 to 32779
Data columns (total 6 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   name     32780 non-null  object 
 1   grape    0 non-null      float64
 2   region   32777 non-null  object 
 3   variety  32422 non-null  object 
 4   rating   32780 non-null  float64
 5   notes    32780 non-null  object 
dtypes: float64(2), object(4)
memory usage: 1.8+ MB


In [16]:
# Sort data
df.sort_values(by = "rating", ascending = False).head()

name  grape  \
9986           Chateau Angelus (Futures Pre-Sale) 2019    NaN   
21597          Espectacle Espectacle del Montsant 2012    NaN   
12857  Chateau Pavie (1.5 Liter Futures Pre-Sale) 2019    NaN   
25936                 Guigal La Turque Cote Rotie 2010    NaN   
12856  Chateau Pavie (1.5 Liter Futures Pre-Sale) 2018    NaN   

                              region   variety  rating  \
9986   St. Emilion, Bordeaux, France  Red Wine    99.0   
21597                          Spain  Red Wine    99.0   
12857  St. Emilion, Bordeaux, France  Red Wine    99.0   
25936      Cote Rotie, Rhone, France  Red Wine    99.0   
12856  St. Emilion, Bordeaux, France  Red Wine    99.0   

                                                   notes  
9986   This 2019 vintage, made while the estate was u...  
21597  Its color is surprisingly intense compared to ...  
12857  Blend: 50% Merlot, 32% Cabernet Franc, 18% Cab...  
25936  La Turque displays deep ruby red color with da...  
12856  Blend: 60% Merlot, 22% Cabernet Franc and 18% ...

In [17]:
# Do some replacements
df = df.replace({"\r": ""}, regex = True)
df = df.replace({"\n": ""}, regex = True)
df.head(10)

name  grape  \
0  1000 Stories Bourbon Barrel Aged Batch Blue Ca...    NaN   
1  1000 Stories Bourbon Barrel Aged Gold Rush Red...    NaN   
2  1000 Stories Bourbon Barrel Aged Gold Rush Red...    NaN   
3    1000 Stories Bourbon Barrel Aged Zinfandel 2013    NaN   
4    1000 Stories Bourbon Barrel Aged Zinfandel 2014    NaN   
5    1000 Stories Bourbon Barrel Aged Zinfandel 2016    NaN   
6    1000 Stories Bourbon Barrel Aged Zinfandel 2017    NaN   
7                            12 Linajes Crianza 2014    NaN   
8                            12 Linajes Reserva 2012    NaN   
9                   14 Hands Cabernet Sauvignon 2010    NaN   

                        region   variety  rating  \
0        Mendocino, California  Red Wine    91.0   
1                   California  Red Wine    89.0   
2                   California  Red Wine    90.0   
3      North Coast, California  Red Wine    91.0   
4                   California  Red Wine    90.0   
5                   California  Red Wine    91.0   
6                   California  Red Wine    92.0   
7      Ribera del Duero, Spain  Red Wine    92.0   
8      Ribera del Duero, Spain  Red Wine    94.0   
9  Columbia Valley, Washington  Red Wine    87.0   

                                               notes  
0  This is a very special, limited release of 100...  
1  The California Gold Rush was a period of coura...  
2  The California Gold Rush was a period of coura...  
3  The wine has a deep, rich purple color. An int...  
4  Batch #004 is the first release of the 2014 vi...  
5  1,000 Stories Bourbon barrel-aged Zinfandel is...  
6  Batch 55 embodies an opulent vintage, which sa...  
7  Red with violet hues. The aromas are very inte...  
8  On the nose, a complex predominance of mineral...  
9  Concentrated aromas of dark stone fruits and t...

In [20]:
# Remove columns
df.drop(['grape'], axis = 1, inplace = True) # Inplace indicate don't create other dataframe, just do it in place
df.describe()

rating
count  32780.000000
mean      91.186608
std        2.190391
min       85.000000
25%       90.000000
50%       91.000000
75%       92.000000
max       99.000000

In [21]:
# Group operations
df.groupby("region").mean()

rating
region                                                       
Abruzzo, Italy                                      89.954545
Aconcagua Valley, Chile                             90.633333
Adelaida District, Paso Robles, Central Coast, ...  92.357143
Adelaide Hills, South Australia, Australia          89.625000
Adelaide, South Australia, Australia                90.000000
...                                                       ...
Yakima Valley, Columbia Valley, Washington          91.141304
Yamhill-Carlton District, Willamette Valley, Or...  91.404762
Yarra Valley, Victoria, Australia                   89.630769
Yorkville Highlands, Mendocino, California          92.000000
Yountville, Napa Valley, California                 93.473684

[404 rows x 1 columns]